src/util.ts

```diff
@@ -127,854 +236,704 @@ class PartialFailureError extends Error {
- * @param {function} callback - The callback function.
- */
-function handleResp(err: Error, resp: r.Response, body: any, callback: BodyResponseCallback) {
-  callback = callback || util.noop;
-
-  const parsedResp = extend(
-    true,
-    {err: err || null},
-    resp && util.parseHttpRespMessage(resp),
-    body && util.parseHttpRespBody(body)
-  );
-
-  callback(parsedResp.err, parsedResp.body, parsedResp.resp);
-}
-
-util.handleResp = handleResp;
-
-/**
- * Sniff an incoming HTTP response message for errors.
- *
- * @param {object} httpRespMessage - An incoming HTTP response message from `request`.
- * @return {object} parsedHttpRespMessage - The parsed response.
- * @param {?error} parsedHttpRespMessage.err - An error detected.
- * @param {object} parsedHttpRespMessage.resp - The original response object.
- */
-function parseHttpRespMessage(httpRespMessage: r.Response) {
-  const parsedHttpRespMessage: any = {
-    resp: httpRespMessage,
-  };
-
-  if (httpRespMessage.statusCode < 200 || httpRespMessage.statusCode > 299) {
-    // Unknown error. Format according to ApiError standard.
-    parsedHttpRespMessage.err = new util.ApiError({
-      errors: [],
-      code: httpRespMessage.statusCode,
-      message: httpRespMessage.statusMessage,
-      response: httpRespMessage,
-    });
-  }
-
-  return parsedHttpRespMessage;
-}
-
-util.parseHttpRespMessage = parseHttpRespMessage;
-
-interface ParsedHttpResponseBody {
-  body: any;
-  err?: Error;
-}
-
-/**
- * Parse the response body from an HTTP request.
- *
- * @param {object} body - The response body.
- * @return {object} parsedHttpRespMessage - The parsed response.
- * @param {?error} parsedHttpRespMessage.err - An error detected.
- * @param {object} parsedHttpRespMessage.body - The original body value provided
- *     will try to be JSON.parse'd. If it's successful, the parsed value will be
- *     returned here, otherwise the original value.
- */
-function parseHttpRespBody(body: any) {
-  const parsedHttpRespBody: ParsedHttpResponseBody = {
-    body,
-  };
-
-  if (is.string(body)) {
-    try {
-      parsedHttpRespBody.body = JSON.parse(body);
-    } catch (err) {
-      parsedHttpRespBody.err = new util.ApiError('Cannot parse JSON response');
-    }
-  }
+export interface MakeRequestConfig {
+  /**
+   * Automatically retry requests if the response is related to rate limits or certain intermittent server errors.
+   * We will exponentially backoff subsequent requests by default. (default: true)
+   */
+  autoRetry?: boolean;
 
-  if (parsedHttpRespBody.body && parsedHttpRespBody.body.error) {
-    // Error from JSON API.
-    parsedHttpRespBody.err = new util.ApiError(parsedHttpRespBody.body.error);
-  }
+  /**
+   * Maximum number of automatic retries attempted before returning the error. (default: 3)
+   */
+  maxRetries?: number;
 
-  return parsedHttpRespBody;
+  stream?: duplexify.Duplexify;
 }
 
-util.parseHttpRespBody = parseHttpRespBody;
+export class Util {
```

Comment:
Initial thought -- I don't know that making `Util` an instantiable object makes sense, since util files are generally just collections of helper functions. I think we should be exporting each function individually instead, so other files only import the functions they need.

Reply:
That's a great question!  This was my initial approach.  When I did that, it broke pretty much all of the tests.  The tests are mocking `util` in a way that depends on the same function reference being stubbed as one invoked by the function under test.  Essentially - unless there's an object instance that can have it's methods stubbed reliably, none of the tests will come close to working.

It's possible I'm missing something, but this felt like a reasonable first step in the transition.

Reply:
You can use `shimmer` to wrap/mock functions directly exported by a module. This will work:
```ts
import * as shimmer from 'shimmer';
import { privatize } from '../src/util';
import * as util from '../src/util';

shimmer.wrap(util, 'privatize', () => {
  return () => 'foo';
});

console.log(privatize({}, '', '')); // prints 'foo'
```
However, it won't work when mocking a function exported by `util` to test a different function exported by `util` (in other words, if `util` exports a function that calls another function in `util`).

Does that work for the cases you described?

Reply:
> However, it won't work when mocking a function exported by util to test a different function exported by util (in other words, if util exports a function that calls another function in util).

This is *exactly* what I was running into, and why I had to use the instance approach :(